Add interface declarations to JIT

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21972 Add interface declarations to JIT**
* #25058 Simplify NamedType

Summary: this is the first commit in a series to add interfaces to JIT.
Interfaces allow the specification through a blank python class of an
abstract interface that can be used in type annotations for Script functions.
If a TorchScript class implements all the methods in the interface with
the appropriate types, then it is implicitly considered to implement
that interface.

Follows required:
* implementation of serialization
* implementation in the parser frontend
* better error reporting for explaining why a class does not meet an
  interface specification.

Differential Revision: [D15901930](https://our.internmc.facebook.com/intern/diff/D15901930)

zdevito committed:
Add interface declarations to JIT

Summary: this is the first commit in a series to add interfaces to JIT.
Interfaces allow the specification through a blank python class of an
abstract interface that can be used in type annotations for Script functions.
If a TorchScript class implements all the methods in the interface with
the appropriate types, then it is implicitly considered to implement
that interface.

Follows required:
* implementation of serialization
* implementation in the parser frontend
* better error reporting for explaining why a class does not meet an
  interface specification.

zdevito commented:
For some reason the previous PR got into a weird 'merged' state, so I had to open a new one.

zdevito committed:
Update on "Add interface declarations to JIT"

Add interface declarations to JIT

Summary: this is the first commit in a series to add interfaces to JIT.
Interfaces allow the specification through a blank python class of an
abstract interface that can be used in type annotations for Script functions.
If a TorchScript class implements all the methods in the interface with
the appropriate types, then it is implicitly considered to implement
that interface.

Follows required:
* implementation of serialization
* implementation in the parser frontend
* better error reporting for explaining why a class does not meet an
  interface specification.

gh-metadata: pytorch pytorch 21972 gh/zdevito/69/head

zdevito committed:
Update on "Add interface declarations to JIT"

Add interface declarations to JIT

Summary: this is the first commit in a series to add interfaces to JIT.
Interfaces allow the specification through a blank python class of an
abstract interface that can be used in type annotations for Script functions.
If a TorchScript class implements all the methods in the interface with
the appropriate types, then it is implicitly considered to implement
that interface.

Follows required:
* implementation of serialization
* implementation in the parser frontend
* better error reporting for explaining why a class does not meet an
  interface specification.

gh-metadata: pytorch pytorch 21972 gh/zdevito/69/head

suo commented on torch/csrc/jit/script/script_type_parser.cpp:
```diff
@@ -198,6 +198,139 @@ TypePtr ScriptTypeParser::parseType(const std::string& str) {
   return parseTypeFromExpr(p.parseExp());
 }
+
+std::vector<IValue> ScriptTypeParser::evaluateDefaults(
```
next time we should split up the mechanical moves from the actual changes 😛

suo commented on torch/jit/__init__.py:
```diff
@@ -1090,6 +1090,12 @@ def _enable_recursive_script():
 
+def _get_class_ast(obj):
+    if not _is_new_style_class(obj):
+        raise RuntimeError("TorchScript classes must be new-style classes. Please inherit from 'object'")
```
Update this message?

suo reviewed (approved):
NGNT

zdevito committed:
Update on "Add interface declarations to JIT"

Add interface declarations to JIT

Summary: this is the first commit in a series to add interfaces to JIT.
Interfaces allow the specification through a blank python class of an
abstract interface that can be used in type annotations for Script functions.
If a TorchScript class implements all the methods in the interface with
the appropriate types, then it is implicitly considered to implement
that interface.

Follows required:
* implementation of serialization
* implementation in the parser frontend
* better error reporting for explaining why a class does not meet an
  interface specification.

gh-metadata: pytorch pytorch 21972 gh/zdevito/69/head

zdevito committed:
Update on "Add interface declarations to JIT"

Add interface declarations to JIT

Summary: this is the first commit in a series to add interfaces to JIT.
Interfaces allow the specification through a blank python class of an
abstract interface that can be used in type annotations for Script functions.
If a TorchScript class implements all the methods in the interface with
the appropriate types, then it is implicitly considered to implement
that interface.

Follows required:
* implementation of serialization
* implementation in the parser frontend
* better error reporting for explaining why a class does not meet an
  interface specification.

gh-metadata: pytorch pytorch 21972 gh/zdevito/69/head
